### PR TITLE
Ping metrics

### PR DESCRIPTION
**Pull request by dmke**
This enables fetching device statistics from UNMS, extracting the ping RTT values from it, and export them back to Prometheus.

### CODE REVIEW

**Comment by dmke** [comment]
I've fixed a unit confusion in my last force-push:

- for historical reasons, `(github.com/digineo/go-ping/monitoring).Result` captures RTT values as float32-encoded millisecond values
- I forgot to take this into account when exporting the `unms_ping_rtt_*_seconds` metrics

Since fixing `github.com/digineo/go-ping` is not easily possible (currently no capacity to do so), I've dropped that dependency and imported the relevant code (found in `exporter/ping_rtt.go`).

**Comment by mraerino** [comment]
the diff looks like it contains things from #22 - maybe rebase just in this case?

**Comment by dmke** [comment]
Hm. Rebasing through the GitHub interface didn't work as expected. I can drop [335571a](https://github.com/ffddorf/unms-exporter/pull/23/commits/335571a7ea16c407d7e5dca6bd4b3f0d710d2080) from this PR, but I'm currently not at my workstation...

**Review by mraerino — COMMENTED** [review]
the implementation looks good to me.

could you make this feature available behind a flag or query parameter though?
i'm not sure all users of this service want to cause these additional requests to the API.

**Comment by dmke** [comment]
Sure, you got some name in mind? Previous iterations used `--with-statistics` and later `--enable-ping-metrics`.

The former name stemmed from the API endpoint (`/devices/{id}/statistics`), but that didn't feel right, because we're only grabbing the ping metrics. The latter does reflect more the actual purpose, but will become outdated should any of the other metrics become interesting (among others, the API provides data for client connection quality, uplink/downlink/wireless utilization, WiFi signal strength, temperature).

Maybe something akin to `--extra-metrics=<comma separated list>` (i.e. `--extra-metrics=ping`), to be safe for future extension?

**Comment by dmke** [comment]
pflag and/or viper seem to have problems mapping `--extra-metrics`:

```go
type config struct {
	...
	ExtraMetrics []string `mapstructure:"extra_metrics" envconfig:"extra_metrics"`
}

func main() {
	...
	flags.StringSlice("extra-metrics", nil, "Enable additional metrics")
	...
	v.BindPFlags(flags)
	v.Unmarshal(conf)
	log.Printf("extra: %+v\n", conf.ExtraMetrics)
	...
}
```

```console
$ UNMS_EXPORTER_EXTRA_METRICS=ping,ccq go run main.go
INFO[0000] extra: [ping ccq]

$ go run main.go --extra-metrics ping,ccq   
INFO[0000] extra: []
```

If I change it to `flags.StringSlice("extra_metrics", ...)`, it works:

```console
$ go run main.go --extra_metrics ping,ccq   
INFO[0000] extra: [ping ccq]
```

But I'm not a fan of underscores in flags. One option could be to just use a single-word flag name (`--extras`). Any other suggestions?

**Comment by dmke** [comment]
That doesn't actually make a difference:

```go
flags.StringSliceVar(&conf.ExtraMetrics, "extra-metrics", nil, "Enable additional metrics")
```
```console
$ go run main.go --extra-metrics ping,ccq
INFO[0000] extra: [ping ccq]

$ UNMS_EXPORTER_EXTRA_METRICS=ping,ccq go run main.go
INFO[0000] extra: []
```

Even with a bit trickery I can't convince Viper to match the environment variable to the CLI flag:

```go
flags.StringSlice("extra-metrics", nil, "Enable additional metrics")
flags.StringSlice("extra_metrics", nil, "Enable additional metrics")
flags.MarkHidden("extra_metrics")
```
```console
$ go run main.go --extra-metrics ping,ccq            
INFO[0000] extra: []

$ UNMS_EXPORTER_EXTRA_METRICS=ping,ccq go run main.go
INFO[0000] extra: [ping ccq]

$ go run main.go --extra_metrics ping,ccq
INFO[0000] extra: [ping ccq]
```

**Comment by dmke** [comment]
> [...] you're confusing [...] `mapstructure` [and] `envconfig`

You're right, I tripped over that.

> setting `envconfig:"extra_metrics"` should solve the case

Already done [here](https://github.com/ffddorf/unms-exporter/pull/23#issuecomment-1096976325). That's not it.

I'd probably need to set `mapstructure:"extra-metrics"`, but that would then introduce an inconsistency with `mapstructure:"log_level"`...

I've settled on a single config word (`extras`), which may not be 100% perfect, but avoids these edge cases.

**Comment by mraerino** [comment]
@dmke i opened a PR against your PR to get the dash option working: https://github.com/digineo/unms-exporter/pull/1

the tests made it a lot easier to implement, thanks for that!

**Comment by mraerino** [comment]
if you want to resolve the conflicts, this might be good to go!

**Comment by dmke** [comment]
Done.

I had to re-introduce the `error`-return for `internal/handler.New()`, I knew it was there for something important :)

File: exporter/exporter.go [hunk 1]
```diff
@@ -241,6 +247,21 @@ func (e *Exporter) collectImpl(ctx context.Context, out chan<- prom.Metric) erro
 			out <- e.newMetric("wan_rx_rate", prom.GaugeValue, wanIF.Statistics.Rxrate, deviceLabels...)
 			out <- e.newMetric("wan_tx_rate", prom.GaugeValue, wanIF.Statistics.Txrate, deviceLabels...)
 		}
+
+		// Ping metrics, if enabled
+		if e.extras.Ping {
```
Comment (mraerino) [inline exporter/exporter.go:252]: do you need to check this here? should be fine to use the nil-check in PingMetrics right? that way the `fetchDeviceData` is the only place it's handled, which i'd like.

**Review by mraerino — COMMENTED** [review]
one thing

File: exporter/exporter.go (hunk 1 above)
Comment (dmke) [reply exporter/exporter.go:252]: I'd rather leave it as is, because future extensions to the `models.DeviceStatistics` struct (hint: coming very soon) will require this conditional where it is now.

**Review by mraerino — COMMENTED** [review]
interesting feature. thanks for the contribution!
